### PR TITLE
Move 'Create Package' modal to its own page

### PR DESCRIPTION
**Pull request by dmarcoux**
With some refactoring... :)

Refer to the commit messages for details.

preview-view

preview-errors

### CODE REVIEW

File: src/api/app/policies/package_policy.rb [hunk 1]
```diff
@@ -1,4 +1,18 @@
 class PackagePolicy < ApplicationPolicy
+  def initialize(user, record, ignore_lock = false)
```
Comment (dmarcoux) [inline src/api/app/policies/package_policy.rb:2]: I could have probably worked around this code smell by [using a context as explained in Pundit's README](https://github.com/varvet/pundit#additional-context), but this would have involved even more refactoring. I decided this was already good enough for this PR.

File: src/api/config/routes/webui_routes.rb [hunk 2]
```diff
@@ -66,7 +66,8 @@
         get 'package/revisions/:project/:package' => :revisions, constraints: cons, as: 'package_view_revisions'
         post 'package/submit_request/:project/:package' => :submit_request, constraints: cons
         get 'package/rdiff/:project/:package' => :rdiff, constraints: cons, as: 'package_rdiff'
-        post 'package/save_new/:project' => :save_new, constraints: cons, as: 'save_new_package'
+        post 'package/create/:project' => :create, constraints: cons, as: 'packages'
```
Comment (dmarcoux) [inline src/api/config/routes/webui_routes.rb:69]: This would be better if we used resources, but it involves a lot more changes due to the huge amount of routes defined here.

File: src/api/app/controllers/source_controller.rb [hunk 3]
```diff
@@ -1103,7 +1103,7 @@ def verify_can_modify_target!
 
     if Package.exists_by_project_and_name(@target_project_name, @target_package_name, follow_project_links: false)
       verify_can_modify_target_package!
-    elsif !@project.is_a?(Project) || !User.session!.can_create_package_in?(@project)
+    elsif !@project.is_a?(Project) || !Pundit.policy(User.session!, Package.new(project: @project)).create?
```
Comment (vpereira) [inline src/api/app/controllers/source_controller.rb:1106]: NITPICK: just wondering if it shouldn't be done in another PR

File: src/api/app/models/bs_request_permission_check.rb [hunk 4]
```diff
@@ -385,10 +385,10 @@ def set_permissions_for_action(action, new_state = nil)
         @write_permission_in_this_action = true
       end
     else
-      if @target_project && User.session!.can_create_package_in?(@target_project, true)
+      if @target_project && Pundit.policy(User.session!, Package.new(project: @target_project), true).create?
```
Comment (vpereira) [inline src/api/app/models/bs_request_permission_check.rb:388]: NITPICK: just wondering if it shouldn't be done in another PR

File: src/api/app/models/package.rb [hunk 5]
```diff
@@ -1015,6 +1016,12 @@ def valid_name
     errors.add(:name, 'is illegal') unless Package.valid_name?(name)
   end
 
+  def is_unique
```
Comment (vpereira) [inline src/api/app/models/package.rb:1019]: why not `unique?`

File: src/api/app/models/user.rb [hunk 6]
```diff
@@ -448,6 +448,7 @@ def has_global_permission?(perm_string)
     end
   end
 
+  # FIXME: This should be a policy
```
Comment (vpereira) [inline src/api/app/models/user.rb:451]: All the TODO here makes me wonder if we couldn't  call the Pundit Policy in the background without dropping those methods to avoid breaking API.

NITPICK: just wondering if it shouldn't be done in another PR

File: src/api/app/policies/package_policy.rb [hunk 7]
```diff
@@ -1,4 +1,17 @@
...
+    super(user, record)
+    @ignore_lock = ignore_lock
+  end
+
+  def create?
+    return false if !@ignore_lock && record.project.is_locked?
+    return true if user.is_admin? ||
```
Comment (vpereira) [inline src/api/app/policies/package_policy.rb:9]: there is specs covering this?

**Comment by dmarcoux** [comment]
@vpereira I agree with your comment `just wondering if it shouldn't be done in another PR`.

I will split the commits into different PRs to make this easier to review.

**Comment by dmarcoux** [comment]
I'll close this PR and split it in multiple PRs. Thanks for the review @vpereira, I'll apply your suggestions in the following PRs.
